libs/pageserver_api/src/models.rs

```diff
@@ -7,18 +7,39 @@ use utils::{
     lsn::Lsn,
 };
 
-/// A state of a tenant in pageserver's memory.
-#[derive(Debug, Clone, Copy, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
 pub enum TenantState {
-    /// Tenant is fully operational, its background jobs might be running or not.
-    Active { background_jobs_running: bool },
-    /// A tenant is recognized by pageserver, but not yet ready to operate:
-    /// e.g. not present locally and being downloaded or being read into memory from the file system.
-    Paused,
-    /// A tenant is recognized by the pageserver, but no longer used for any operations, as failed to get activated.
+    // This tenant is being loaded from local disk
+    Loading,
+
+    // This tenant is being downloaded from cloud storage.
+    Attaching,
+
+    // This tenant exists on local disk, and the layer map has been loaded into memory.
+    // The local disk might have some newer files that don't exist in cloud storage yet.
+    Active,
+
+    // This tenant exists on local disk, and the layer map has been loaded into memory.
```

Comment:
> There's a new function, get_active_tenant, that includes the check. get_tenant now returns the tenant regardless of the state it's in. There are comment at those functions that explain this.

Have checked that and for now have a feeling that `get_tenant` could be a private function + a few extra functions to return timeline and tenant infos. But will have to look more and comment on that later.

> ... and only checked when we try to stop the tenant already?
> Attaching and Loading do the same role in the checks and not checked during get_tenant, again

I think, it's me overlooking `get_tenant` and `get_active_tenant` case while searching for usages of particular enum variant usages instead: there's from none to close to none real logic in the code that distinguishes types `TenantState::Attaching`, `TenantState::Loading`, `TenantState::Attaching`, `TenantState::Stopping` and `TenantState::Broken`, we use them more like a documentation and logging tool.

I suspect, this might not stay this way for all enum variants: `Broken` need some alerts/healthchecks, etc.; while other variants will keep more or less the same logic behind them (`TenantState::Loading` and `TenantState::Attaching`, I imagine: we can only wait for that status to change or shutdown).
Instead of having such subtle mixture of docs and ideas, I would try to actually split tenant (and, later, timeline) states on bare minimum of enum variants (aka states with big transitions between them), so that we don't have a combinatorical explosion when dealing with "state A -> state B" transition logic.

As bare minimum, I see the possibility to read and write into timeline, hence trying to stick to enum variants, that could be defined as 

In particular, I see original `TenantState` close to a good one
```rust
/// A state of a tenant in pageserver's memory.
#[derive(Debug, Clone, Copy, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
pub enum TenantState {
    /// r + w
    Active { background_jobs_running: bool },
    /// r? + !w
    /// ?? potentially, we need to split more here, but should it be another enum variant? feels like a field to have instead, `reason: String`
    Paused,
    /// !r + !w; irreversible
    Broken,
}
```

when I start to map new variants to the same idea, I feel like it gets more complex already, knowing that code refactoring could change those guarantees later without touching the enum:
```rust
pub enum TenantState {
    // !r + !w ; should happen soon (since it's from disk, not remote storage)
    Loading,

    // !r + !w; should happen sometimes later, 
    // unknown when due to being downloaded from the network
    Attaching,

    // r + w
    Active,

    // !r + !w; irreversible but presumably can degrade to `Broken`
    Stopping,

    // !r + !w; irreversible
    Broken,
}
```

I would rather concentrate on basic tenant and timeline cases covered first and then publishing another status in a PR with some motivation behind it.

> They are also quite confusing, since downloading = put files on disk + attach (same as we do on start).

When I see `Attaching`, I'm instantly thinking of https://github.com/neondatabase/neon/blob/6ace79345de41e4569856c7ad6978cb1bb1e1765/pageserver/src/tenant_mgr.rs#L117 function that's used for both `TenantState::Loading ` and `TenantState::Attaching` procedures (the latter, remote storage download, downloads blobs on disk and then does the same `attach_local_tenants` on the downloaded files).
But `Attaching` in this context is remote storage download + that attach function call, which is a way longer story.